src/Adapter/Configuration.php

```diff
@@ -102,14 +105,47 @@ public function get($key, $default = null)
             return constant($key);
         }
 
+        return $this->doGet($key, $default);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getForShop(string $key, ShopConstraint $shopConstraint)
```

Comment:
Side note: why introduce a `ShopConstraint` object? Because just providing shopID or just shopGroupId separately won't do the job. There is a fallback behaviour when dealing with multishop configurations because they are not necessarily overridden so the priority is:

1. shopId
2. groupShopId
3. global

Which is why I introduced this object that include both `ShopId` and `ShopGroupId` for conveniency (one parameter)

It could be reshaped as `public function getForShop(string $key, ShopId $shopId, ShopGroupId $shopGroupId)` but then we would need to make some parameters optional..

Reply:
Also one drawback in this behaviour is that the fallback system is automatic, meaning you can't get the value for this specific Shop and get null or an exception when it doesn't exist, you will get the fallback value but you won't know if it comes from the Shop, its Group or the Global...

Reply:
I started introducing a `ConfigurationValue` VO but stopped to avoid over-engineering this PR, the idea was to store the key and the value and add a few getters (getInt, getBool, getByLang, ...) for conveniency.

It could be used both for input and output, and it could also store the "source" shopId or shopGroupID, but this is far beyond the scope of this PR imo, the problem being that we start integrating interfaces here that are not mature enough, because of lack of time and discussion..

Reply:
`ShopConstraint` makes sense. 

To control fallback you can add an optional parameter `$fallback = true`.

That said, I think we need a better `Configuration` class because multishop is a core feature which is incompatible with a key value storage like ParameterBag (which this class extends). So trying to put it inside this class is basically creating a Frankenstein for the sake of not introducing a new class.

Reply:
Yep and there are even some methods of the ParameterBag that are overridden to thrown exception because they are not handled.. So I agree that this Adapter class is only temporary and should be completely rebuilt, but not in this PR
The fallback principle makes sense even as a core feature I guess 🤔. So I can still implement it now even with the Adapter implementation

Reply:
Although it won't be used right now, so it might be safer to wait for next version to add this feature To make sure we make it right